tests/aws/services/lambda_/test_lambda_api.py

```diff
@@ -5294,6 +5294,101 @@ def check_esm_active():
+            QueueUrl=destination_queue_url, AttributeNames=["QueueArn"]
+        )["Attributes"]["QueueArn"]
+
+        dynamodb_create_table(table_name=table_name, partition_key="id")
+        _await_dynamodb_table_active(aws_client.dynamodb, table_name)
+        update_table_response = aws_client.dynamodb.update_table(
+            TableName=table_name,
+            StreamSpecification={"StreamEnabled": True, "StreamViewType": "NEW_IMAGE"},
+        )
+        snapshot.match("update_table_response", update_table_response)
+        stream_arn = update_table_response["TableDescription"]["LatestStreamArn"]
+
+        create_lambda_function(
+            handler_file=TEST_LAMBDA_PYTHON_ECHO,
+            func_name=function_name,
+            runtime=Runtime.python3_12,
+            role=lambda_su_role,
+        )
+        # "minimal"
+        create_response = aws_client.lambda_.create_event_source_mapping(
+            FunctionName=function_name,
+            EventSourceArn=stream_arn,
+            DestinationConfig={"OnFailure": {"Destination": destination_queue_arn}},
+            BatchSize=1,
+            StartingPosition="TRIM_HORIZON",
+            MaximumBatchingWindowInSeconds=1,
+            MaximumRetryAttempts=1,
+        )
+
+        uuid = create_response["UUID"]
+        cleanups.append(lambda: aws_client.lambda_.delete_event_source_mapping(UUID=uuid))
+        snapshot.match("create_response", create_response)
+
+        # the stream might not be active immediately(!)
+        def check_esm_active():
+            return aws_client.lambda_.get_event_source_mapping(UUID=uuid)["State"] != "Creating"
+
+        assert wait_until(check_esm_active)
+
+        get_response = aws_client.lambda_.get_event_source_mapping(UUID=uuid)
+        snapshot.match("get_response", get_response)
+
+        delete_function_response = aws_client.lambda_.delete_function(FunctionName=function_name)
+        snapshot.match("delete_function_response", delete_function_response)
+
+        # LocalStack: Due to ESM deletions being asynchronous, we need to test that the ESM tears
+        # down correctly when the target Lambda has been deleted.
+        if not is_aws_cloud() and is_v2_esm():
+            from localstack.services.lambda_.event_source_mapping.esm_worker import EsmWorker
+
+            def _assert_function_deleted():
+                with pytest.raises(aws_client.lambda_.exceptions.ResourceNotFoundException):
+                    aws_client.lambda_.get_function(FunctionName=function_name)
+                return True
+
+            def _delete_esm_in_store_with_delay(self):
+                wait_until(_assert_function_deleted)
+                self.delete_esm_in_store()
+
+            monkeypatch.setattr(EsmWorker, "delete_esm_in_store", _delete_esm_in_store_with_delay)
+
+        get_response_post_delete = aws_client.lambda_.get_event_source_mapping(UUID=uuid)
+        snapshot.match("get_response_post_delete", get_response_post_delete)
```

Comment:
FYI: Here we get the event source mapping AFTER a lambda has been deleted

Reply:
Would be interesting to see if the state would change if we would trigger some events, but the function is deleted. Not in scope for this PR though!